### ISSUE TITLE
merge-base algorithm improvements 

### ISSUE THREAD

**Issue by grimmySwe** [ingydotnet/git-subrepo#130]
Current implementation of the merge-base is a simple loop of loops. With a subrepo S and a local copy of that C:
best case: S+C (We always go through all commits at least once to detect duplicates)
worst case: S*C + S+C (Find a match and also go through to find duplicates)

As C and S grows the time spent finding merge base will increase rapidly. You could though argue that the normal case should often be more like the best case because if you pull/push regularly you will find matches sooner.

But... worst case is having a common clone commit and then have a small change in you C that you never push.

Implementation suggestion for the search:
Go through C, store all commits in a set based on tree hash, in each entry store a list of C commits and a reference to the parent tree hash.
Go through S, store all commits in the same set, in each entry store a list of S commits.

Final step is to start at the last tree hash and iterate through them by using the parent references to next tree hash:
If you find an entry with exactly one C commit and one S commit you have a unique spot that can be used as merge base.
Entries with more then one C or S can't be used as they indicate that things have been reverted and we can't determine the actual order.

Not sure how to do this in bash though :-)

**Comment by ingydotnet** [comment]
A general thought... If this is just a speed optimization, it doesn't need to
be bash.

ie You could write it in Perl or Python and use the optimization when the
user's environment was suitable.

I do this in `ingydotnet/git-hub`. That tool has a Bash JSON parser, which can
be slow to parse an array of 100 object with about 40 key-pairs each (a common
list command payload).

If the user's env has Perl with JSON module (fairly common by default) then it
will use that.

Both algorithms need to be exactly the same API/behaviour (obviously).

**Comment by ingydotnet** [comment]
It would be neat to script a simulation using a massive existing public Git
repo out there, to see how bad things can be.

Basically your script can `subrepo clone` a commit from 10 years ago and do a
pull 5 years later (or somesuch).

**Comment by AndrewSmart** [comment]
Ran into this S*C merge-base & large history performance considerations a year ago. Deleted that branch after Ingy's remix, I regret that now.
I recall that in the lookup of the tree-hashes /commits something was meaningfully faster than git log (git rev-list?). I can see git log is used currently. I also looked into shrinking the lists some (I'm guessing `git rev-list --remove-empty`?). Have no clue what I did to shrink them, or any of the other magic.
Looking at the worst case considering how large repos can get lowered my confidence in my approach :-/ Neat to see you ran into the same.

In short, don't forget to look into shrinking S and C as well as speeding up their construction :-)

Anyway, excellent work, thank you!

**Comment by grimmySwe** [comment]
Regarding bash... I found that there is hash maps in bash and that would come a long way:
http://stackoverflow.com/questions/1494178/how-to-define-hash-tables-in-bash

With the suggested algorithm above I can make the Worst case to S+C and I can't see anyway to improve further as you want to go through all at least once to check for duplicates.

**Comment by ingydotnet** [comment]
Be careful. We need to support Bash 3.2 forever thanks to OSX. :(

If hashes are a big win though, I'll accept a dual 3.2/4.0 solution. ie
faster if you don't use a mac :)

**Comment by grimmySwe** [comment]
I am testing a little hack variant:

```
#!/bin/bash

git:log-tree() {
    git log --pretty=format:%H:%T "$1"
}

prepare() {
    local prefix="$1";
    local data=$(git:log-tree "$2");

    for entry in $data; do
        local array=(${entry//:/ })
        local commit=${array[0]}
        local tree=${array[1]}

        local key="${prefix}_$tree";

        if [ -z ${!key+x} ]; then
            declare -g "$key=$commit";
            #echo "Declare $key";
        else
            # Value is already set!
            # we should do a reset
            declare -g "$key=${!key}_$commit";
            #echo "$key is set to '${!key}'";
        fi
        if $3; then
            order+=("$tree");
        fi
    done

}

merge-base() {
    local order=();

    prepare "left" "$1" true;
    prepare "right" "$2" false;

    for entry in ${order[@]}; do
        local left_key="left_$entry";
        local left_value="${!left_key}";
        if [[ ${#left_value} > 40 ]]; then
            echo "Duplicate_left: $left_value";
        else
            local right_key="right_$entry";
            if [ -z ${!right_key+x} ]; then
                echo "Not found in right";
            else
                local right_value=${!right_key};
                if [[ ${#right_value} > 40 ]]; then
                    echo "Duplicate_right $right_value";
                else
                    echo "Found ancestor: ${!left_key},${!right_key}";
                fi
            fi
        fi
    done
}

merge-base "$1" "$2";
```

Instead of using a real hash/dict I use the built in variable mechanism. Hashes have the great benefit of being [0-9a-e]. So in the algorithm above you go through all commits on both side once to create the data and then you iterate through that data until you find the first candidate with equal tree hash and no duplicates.

So the best case is S+C and worst case 2S+C. No nested loops. There are some simplifications that can be done if you disregard debugging, You don't actually need to store the duplicate entires, you could simply mark it as duplicate.

Note 1: I haven't tested this live yet, only on my small repo at home.
Note 2: that the `declare -g`is probably a late bash as well, but it should be possible to work around with keeping everything in a single method or exporting or something else to make the data "global"

**Comment by ingydotnet** [comment]
@grimmySwe yeah I was going to suggest faking hashes, but I figured you'd figure it out.

We should get some real test cases based on real (big) repos. Since you can clone and pull to specific commits way in the past, this should be trivial. I'll work on one for you, unless you already have one.

**Comment by ingydotnet** [comment]
I cloned https://github.com/git/git and ran this command:

```
~/src/git $ time git subrepo merge-base e83c5163316f89bfbde7d9ab23ca2e25604af290 master
Found common ancestor between these commits:
  'e83c5163316f89bfbde7d9ab23ca2e25604af290':HEAD 'e83c5163316f89bfbde7d9ab23ca2e25604af290'
  'master': 'e83c5163316f89bfbde7d9ab23ca2e25604af290'

real	0m2.322s
user	0m2.200s
sys	0m0.181s
```

That's 70273 commits. `e83c516` is the first commit in the repo. 2 secs doesn't seem bad. Am I missing something?

**Comment by grimmySwe** [comment]
What you tested is the best case, you will actually only go through S+C as you find the ancestor in the first attempt.

I tested with finding a commit that is 295 commits back (maint branch)
```
$ time git subrepo merge-base master 1b0b6dd
Found common ancestor between these commits:
  'master': '1b0b6dd0720572dcf90c264aeb91f96a017b0f25'
  '1b0b6dd':HEAD '1b0b6dd0720572dcf90c264aeb91f96a017b0f25'

real    3m11.928s
user    3m9.527s
sys     0m2.704s
```

**Comment by grimmySwe** [comment]
First one thing to show the unbalance in timing, if I do the reverse search it goes really quick:
```
$ time git subrepo merge-base origin/maint master
Found common ancestor between these commits:
  '1b0b6dd':HEAD '1b0b6dd0720572dcf90c264aeb91f96a017b0f25'
  'master': '1b0b6dd0720572dcf90c264aeb91f96a017b0f25'

real    0m3.714s
user    0m3.649s
sys     0m0.200s
```


Testing my new script on the two combinations:
```
$ time ../../bash/mergebase.sh origin/next origin/maint
left 42049
right 41635
Step back: 265

real    0m23.160s
user    0m23.118s
sys     0m0.228s

$ time ../../bash/mergebase.sh origin/maint origin/next
left 41635
right 42049
Found ancestor: 1b0b6dd0720572dcf90c264aeb91f96a017b0f25,1b0b6dd0720572dcf90c264aeb91f96a017b0f25
Step back: 1

real    0m23.058s
user    0m22.984s
sys     0m0.255s
```
Now they take almost the same time. Compare this with the exact same searches in merge-base:
```
]$ time git subrepo merge-base origin/maint origin/next
Found common ancestor between these commits:
  'origin/maint':HEAD '1b0b6dd0720572dcf90c264aeb91f96a017b0f25'
  'origin/next': '1b0b6dd0720572dcf90c264aeb91f96a017b0f25'

real    0m3.671s
user    0m3.590s
sys     0m0.234s
$ time git subrepo merge-base origin/next origin/maint
Found common ancestor between these commits:
  'origin/next': '1b0b6dd0720572dcf90c264aeb91f96a017b0f25'
  'origin/maint':HEAD '1b0b6dd0720572dcf90c264aeb91f96a017b0f25'

real    5m42.411s
user    5m38.138s
sys     0m4.656s
```

**Comment by grimmySwe** [comment]
Updated script to use ```eval``` instead of ```declare -g```. Also added some debugging info on how many entries we actually visit.

```
#!/bin/bash

git:log-tree() {
    git log --pretty=format:%H:%T "$1"
}

prepare() {
    local prefix="$1";
    local data=$(git:log-tree "$2");
    local counter=0;

    for entry in $data; do
        local array=(${entry//:/ })
        local commit=${array[0]}
        local tree=${array[1]}

        local key="${prefix}_$tree";

        if [ -z ${!key+x} ]; then
            # declare -g "$key=$commit";
            eval "$key=$commit";
        else
            # Value is already set!
            # we should do a reset
            # declare -g "$key=${!key}_$commit";
            eval "$key=${!key}_$commit";
        fi
        if $3; then
            order+=("$tree");
        fi
        ((counter++))
    done
    echo $prefix $counter;
}

merge-base() {
    local order=();
    local counter=0;

    prepare "left" "$1" true;
    prepare "right" "$2" false;

    for entry in ${order[@]}; do
        ((counter++));
        local left_key="left_$entry";
        local left_value="${!left_key}";
        if [[ ${#left_value} > 40 ]]; then
            echo "Duplicate_left: $left_value";
        else
            local right_key="right_$entry";
            if [ -z ${!right_key+x} ]; then
                :
                #echo "Not found in right";
            else
                local right_value=${!right_key};
                if [[ ${#right_value} > 40 ]]; then
                    echo "Duplicate_right $right_value";
                else
                    echo "Found ancestor: ${!left_key},${!right_key}";
                    echo "Step back: $counter";
                    return
                fi
            fi
        fi
    done
    echo "No ancestor found";
    echo "Step back: $counter";
}

merge-base "$1" "$2";
```
Seems to be a tiny bit slower but then it should work on OS X bash 3.2 as well.
```
$ time ../../bash/mergebase.sh origin/next origin/maint
left 42049
right 41635
...
Found ancestor: 1b0b6dd0720572dcf90c264aeb91f96a017b0f25,1b0b6dd0720572dcf90c264aeb91f96a017b0f25
Step back: 265

real    0m25.063s
user    0m24.997s
sys     0m0.243s
$ time ../../bash/mergebase.sh origin/maint origin/next
left 41635
right 42049
Found ancestor: 1b0b6dd0720572dcf90c264aeb91f96a017b0f25,1b0b6dd0720572dcf90c264aeb91f96a017b0f25
Step back: 1

real    0m25.248s
user    0m25.187s
sys     0m0.246s
```

**Comment by grimmySwe** [comment]
Note that currently one of the common use cases of simply pulling in a repo will end up slow as there are new commits in S. So for each commit in S not in C you iterate. So say that you have 100 new commits in S, you will iterate through all commits in C 100 times to find the common ancestor.

**Comment by grimmySwe** [comment]
Created a pull request #138 for this.

I am though a little worried about speed for real life repositories, considering that you perform filter-branch operations on the entire history. It might be possible to store the latest common ancestor pair and reuse that as a limit.

**Comment by grimmySwe** [comment]
The cache part is moved to #140 
Closing this.
